### DISCUSSION TITLE
INSERT query with sub-SELECT in VALUES -- Unhandled expression type

### DISCUSSION THREAD

**Discussion by treesloth** [MarkMpn/Sql4Cds#163] [Q&A] [answered]
I'm not sure if this is really a feasible or not.  I have a query like this:

```
INSERT  INTO cstm_saleslead (cstm_salesleadnumber, cstm_accountid, cstm_leadclosedate)
VALUES		('ABC123',
		(SELECT cstm_accountid
		FROM   cstm_account
		WHERE  name = 'Secret Spy Gadgets, Inc'),
		'2023-03-30 13:00:00.000');
```

I get the response:

`Unhandled expression type: (SELECT cstm_accountid FROM   cstm_account WHERE  name = 'Secret Spy Gadgets, Inc')
`

I suppose I might be using a structure that XrmToolBox just doesn't understand, but similar structures with UPDATE, SELECT, etc work perfectly well.  Do I have something wrong here, or am I simply outside the part of SQL that is implemented for SQL 4 CDS?

**Comment by MarkMpn** [comment, marked as answer]
Yes, at the moment subqueries are not supported in the `VALUES` clause of an `INSERT` statement. I've created issue #164 to track this for a future improvement.

For the moment however you could rewrite this query as:

```sql
INSERT  INTO cstm_saleslead (cstm_salesleadnumber, cstm_accountid, cstm_leadclosedate)
SELECT 	'ABC123',
		cstm_accountid,
		'2023-03-30 13:00:00.000'
FROM   cstm_account
WHERE  name = 'Secret Spy Gadgets, Inc'
```

This has slightly different semantics to the subquery version though as it will succeed if there are multiple records with a matching name, whereas the subquery version would produce an error.
